File: apps/ios/MagicMobile/BoardZoneReference.swift
import Foundation

/// Resolves membership from the current authorized snapshot, never a cached card list.
enum BoardZoneReference: Hashable {
    enum PlayerZone: String, CaseIterable {
        case library, hand, battlefield, graveyard, exile, command, stack
    }

    enum NamedKind: String, CaseIterable {
        case exile, companion, revealed, lookedAt

        fileprivate var title: String {
            self == .lookedAt ? "Looked at" : rawValue.capitalized
        }

        fileprivate func groups(in snapshot: GameSnapshot) -> [XmageNamedZone] {
            guard let xmage = snapshot.xmage else { return [] }
            switch self {
            case .exile: return xmage.exileZones
            case .companion: return xmage.companion
            case .revealed: return xmage.revealed
            case .lookedAt: return xmage.lookedAt
            }
        }
    }

    case player(playerID: String, zone: PlayerZone)
    case named(kind: NamedKind, id: String)
    case collection(NamedKind)

    func cards(in snapshot: GameSnapshot) -> [ZoneCard] {
        switch self {
        case let .player(playerID, zone):
            guard let zones = snapshot.players.first(where: { $0.playerId == playerID })?.zones else { return [] }
            switch zone {
            case .library: return zones.library
            case .hand: return zones.hand
            case .battlefield: return zones.battlefield
            case .graveyard: return zones.graveyard
            case .exile: return zones.exile
            case .command: return zones.command
            case .stack: return zones.stack
            }
        case let .named(kind, id):
            return kind.groups(in: snapshot).first { $0.id == id }?.cards ?? []
        case let .collection(kind):
            return kind.groups(in: snapshot).flatMap(\.cards)
        }
    }

    func title(in snapshot: GameSnapshot) -> String {
        switch self {
        case let .player(playerID, zone):
            let name = snapshot.players.first { $0.playerId == playerID }?.displayName
            let label = name.map { EngineDisplayText.label($0) } ?? ""
            let zoneTitle = zone.rawValue.capitalized + ([PlayerZone.hand, .library].contains(zone) ? " — visible cards" : "")
            return label.isEmpty ? zoneTitle : "\(label) · \(zoneTitle)"
        case let .named(kind, id):
            let name = kind.groups(in: snapshot).first { $0.id == id }?.name
            return name.map { EngineDisplayText.label($0, fallback: kind.title) } ?? kind.title
        case let .collection(kind):
            return kind.title
        }
    }

    static func namedReferences(in snapshot: GameSnapshot) -> [BoardZoneReference] {
        NamedKind.allCases.flatMap { kind in
            kind.groups(in: snapshot).map { .named(kind: kind, id: $0.id) }
        }
    }
}

File: apps/android/core/src/main/kotlin/io/magicmobile/android/game/BoardZoneReference.kt
package io.magicmobile.android.game

/** Port of BoardZoneReference.swift. Resolves membership from the current snapshot, never a cached card list. */
sealed class BoardZoneReference {
    enum class PlayerZone(val rawValue: String) { LIBRARY("library"), HAND("hand"), BATTLEFIELD("battlefield"), GRAVEYARD("graveyard"),
        EXILE("exile"), COMMAND("command"), STACK("stack") }

    enum class NamedKind(val rawValue: String) {
        EXILE("exile"), COMPANION("companion"), REVEALED("revealed"), LOOKED_AT("lookedAt");
        val title: String get() = if (this == LOOKED_AT) "Looked at" else capitalizedWords(rawValue)
        fun groups(snapshot: GameSnapshot): List<XmageNamedZone> {
            val x = snapshot.xmage ?: return emptyList()
            return when (this) { EXILE -> x.exileZones; COMPANION -> x.companion; REVEALED -> x.revealed; LOOKED_AT -> x.lookedAt }
        }
    }

    data class Player(val playerID: String, val zone: PlayerZone) : BoardZoneReference()
    data class Named(val kind: NamedKind, val id: String) : BoardZoneReference()
    data class Collection(val kind: NamedKind) : BoardZoneReference()
    data class PlayerEnchantments(val playerID: String) : BoardZoneReference()

    fun cards(snapshot: GameSnapshot): List<ZoneCard> = when (this) {
        is PlayerEnchantments -> ZoneCard.enchanting(playerID, snapshot.players.flatMap { it.zones.battlefield })
        is Player -> snapshot.players.firstOrNull { it.playerId == playerID }?.zones?.let { z ->
            when (zone) {
                PlayerZone.LIBRARY -> z.library; PlayerZone.HAND -> z.hand; PlayerZone.BATTLEFIELD -> z.battlefield
                PlayerZone.GRAVEYARD -> z.graveyard; PlayerZone.EXILE -> z.exile; PlayerZone.COMMAND -> z.command; PlayerZone.STACK -> z.stack
            }
        } ?: emptyList()
        is Named -> kind.groups(snapshot).firstOrNull { it.id == id }?.cards ?: emptyList()
        is Collection -> kind.groups(snapshot).flatMap { it.cards }
    }

    fun title(snapshot: GameSnapshot): String = when (this) {
        is PlayerEnchantments -> "Enchanting ${snapshot.playerLabel(playerID)}"
        is Player -> {
            val label = snapshot.players.firstOrNull { it.playerId == playerID }?.displayName?.let { EngineDisplayText.label(it) } ?: ""
            val zoneTitle = capitalizedWords(zone.rawValue) + if (zone == PlayerZone.HAND || zone == PlayerZone.LIBRARY) " — visible cards" else ""
            if (label.isEmpty()) zoneTitle else "$label · $zoneTitle"
        }
        is Named -> kind.groups(snapshot).firstOrNull { it.id == id }?.name?.let { EngineDisplayText.label(it, kind.title) } ?: kind.title
        is Collection -> kind.title
    }

    companion object {
        fun namedReferences(snapshot: GameSnapshot): List<BoardZoneReference> =
            NamedKind.entries.flatMap { kind -> kind.groups(snapshot).map { Named(kind, it.id) } }
    }
}
